akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -1088,44 +1090,71 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
 
     private def sendEvent(event: T, promise: Promise[Done]): Promise[Done] = {
       onAsyncInput(event, OptionVal.Some(promise))
-      currentState.get() match {
-        case Completed ⇒ failPromiseOnComplete(promise)
-        case _         ⇒ promise
-      }
+      if (stopped) failPromiseOnComplete(promise)
+      else promise
     }
 
     // external call
     override def invokeWithFeedback(event: T): Future[Done] = {
       val promise: Promise[Done] = Promise[Done]()
-      promise.future.andThen { case _ ⇒ waitingForProcessing.remove(promise) }(akka.dispatch.ExecutionContexts.sameThreadExecutionContext)
-      waitingForProcessing.add(promise)
-      invokeWithPromise(event, promise).future
+      promise.future.onComplete(_ ⇒ onFeedbackCompleted(promise))(ExecutionContexts.sameThreadExecutionContext)
+
+      @tailrec
+      def addToWaiting(): Boolean = {
+        val previous = asyncCallbacksInProgress.get()
+        if (previous != null) {
+          val updated = previous + (this -> (previous(this) + promise))
+          if (!asyncCallbacksInProgress.compareAndSet(previous, updated)) addToWaiting()
+          else true
+        } else {
+          false
+        }
+      }
+
+      if (addToWaiting())
+        invokeWithPromise(event, promise).future
+      else
+        Future.failed(streamDetatchedException)
     }
 
-    private def invokeWithPromise(event: T, promise: Promise[Done]): Promise[Done] = {
+    // removes the promise from the callbacks in promise on complete, called from onComplete
+    private def onFeedbackCompleted(promise: Promise[Done]): Unit = {
+      @tailrec
+      def removeFromWaiting(): Unit = {
+        val previous = asyncCallbacksInProgress.get()
+        if (previous != null) {
+          val newSet = previous(this) - promise
+          val updated =
+            if (newSet.isEmpty) previous - this // no outstanding promises, remove stage from map to avoid leak
+            else previous + (this -> newSet)
```

Comment:
if this is in hot path it might be better to use `previous.updated(this, newSet)` to avoid the tuple allocation